geth-utils/gethutil/mpt/trie/stacktrie.go

```diff
@@ -628,7 +635,7 @@ func (st *StackTrie) UpdateAndGetProofs(db ethdb.KeyValueReader, list types.Deri
 	// order that `list` provides hashes in. This insertion sequence ensures that the
 	// order is correct.
 	var indexBuf []byte
-	for i := 1; i < list.Len() && i <= 0x7f; i++ {
+	for i := 0; i < list.Len() && i <= 0x7f; i++ {
```

Comment:
Hi @miha-stopar , have a question regarding the starting index.
Why doest the for loop start from 1 instead of 0? I saw the case of index 0 was handled in L649-L654. But there is only `Update` being called, no `GetProof`. Wondering what will happen if I change to start from 0 and comment out the L649-L654? There is also a simliar case in L655, there is no `GetProof` after index is `0x80`. Wondering the reason why separate into many parts.

Reply:
Yeah, good question :). I tried to remember, but can't say for sure. I guess there was some bug in GetProof and I was debugging it. I think the index can start with 0 and we can delete these two if blocks.

Reply:
What if we could encode `k` in another way so we can get the value we expected? For example, for `i = 0`, we get `k = [0 0 16]`. If so, do we still need the above two blocks?

Another question is I added above two blocks back and in 129 txs case, `k` became `[8 1 8 0 16]` when `i = 129`. Does it fit our expecation? I thought the `k` is increasing like this, `[0 0 16], [0 1 16]... [1 0 16]... [1 15 16], [2 0 16]... [15 15 16], [1 0 0 16]...` dosen't it?

Reply:
For `i = 128`, we have `key = [8, 1, 8, 0]`, that means it's inserted in the branch at position 8 (which is a branch that has the transaction with `i = 0` at the position 0 - it was added in the block `if list.Len() > 0`). The second nibble is 1, so it's inserted at position 1 in the branch.

The state after `i = 128` is:
- top level is a branch with branches at positions 0 - 7 (all these branches are fully filled),
- at position 8 in a top level branch we have a branch with two children at positions 0 (`i = 0`) and 1 (`i = 128`).

Reply:
Thanks for detailed explanation again and I think I know how it works 🤞